src/coreclr/vm/dllimport.cpp

```diff
@@ -2071,6 +2092,10 @@ void NDirectStubLinker::End(DWORD dwStubFlags)
     }
 }
 
+#if defined(TARGET_WINDOWS) && defined(TARGET_X86)
```

Comment:
Why is some of this ifdefed out for Windows x86, and some this only for x86?

Reply:
Due to how we have the copy-constructor marshaller defined, it's less disruptive to the code to not add `TARGET_WINDOWS` checking everywhere and just use `TARGET_X86` for the general feature.

Since we're likely to backport this fix to .NET 8, I wanted to keep the changes a little less disruptive to existing code. Basically, the marshaller changes are guarded by `TARGET_X86` and the actual thunk (assembly and the callback) is guarded by `TARGET_WINDOWS` and `TARGET_X86`, as the copy constructor marshaller exists in CoreCLR on non-Windows platforms today (though triggering it requires custom IL to define the well-known member names).

If we wanted to unify the checks, we have two paths:

- Write the little assembly stub for Linux_x86 as well to maintain parity.
- `#ifdef` out the marshaller on non-Windows, which would technically be a back-compat break.

I'll look at writing the assembly stub for Linux_x86 as it should be very simple and then everything can be `TARGET_X86`.

Reply:
> #ifdef out the marshaller on non-Windows, which would technically be a back-compat break.

How is this a visible back-compat break? Is the marshaller usable in any way on non-Windows?

Reply:
Yes, today if you use IL to write your type with the correctly named `<MarshalCopy>` method and a P/Invoke with the right modopts, we'll use this marshaller. The `Interop/PInvoke/Miscallaneous/CopyCtor` test runs on all CoreCLR platforms and passes today.

Reply:
I do not think it is desirable for these IJW features to work outside Windows.

Reply:
I agree, but I'm very hesitant to remove it for the backport as I've broken customers by removing thought-to-be-unused features before. We can follow up this PR with a full removal of the copy-constructor marshaler on non-Windows, but I don't want to do it in this PR.

Reply:
> I've broken customers by removing thought-to-be-unused features before

This is not a supported scenario in any capacity and requires writing in IL directly. I appreciate the concern here but it is unlikely anyone relying on this is something we would accept as valid.